python/shotgun_model/shotgun_query_model.py

```diff
@@ -0,0 +1,979 @@
+
+        :param sg_data: Shotgun data dictionary that was received from Shotgun.
+        """
+        # default implementation does nothing
+        pass
+
+    def _set_tooltip(self, item, data):
+        """
+        Sets the tooltip for the supplied item.
+
+        Called when an item is created.
+
+        :param item: Shotgun model item that requires a tooltip.
+        :param data: Dictionary of the SG data associated with the model.
+        """
+        # the default implementation does not set a tooltip
+        pass
+
+    ############################################################################
+    # protected convenience methods. these methods can be used by subclasses
+    # to manipulate and manage data returned from Shotgun.
+
+    def _do_depth_first_tree_deletion(self, node):
+        """
+        Depth first interation and deletion of all child nodes
+
+        :param node: :class:`~PySide.QtGui.QStandardItem` tree node
+        """
+
+        # depth first traversal
+        for index in xrange(node.rowCount()):
+            child_node = node.child(index)
+            self._do_depth_first_tree_deletion(child_node)
+
+        # delete the child leaves
+        for index in range(node.rowCount())[::-1]:
+            node.removeRow(index)
+
+    def _get_all_item_unique_ids(self):
+        """
+        Conveneince method. Returns the unique IDs of all items in the model.
+
+        The unique IDs correspond to the field defined by
+        ``SG_DATA_UNIQUE_ID_FIELD``
+
+        :return: A list of uniqe ids for all items in the model.
+        :rtype: ``list``
+        """
+
+        return self._tree_data.keys()
+
+    def _get_item_by_unique_id(self, uid):
+        """
+        Convenience method. Returns an item given a unique ID.
+
+        The unique ``uid`` corresponds to the field defined by
+        ``SG_DATA_UNIQUE_ID_FIELD``
+
+        :param id: The unique id for an item in the model.
+
+        :return: An item corresponding to the supplied uniqueid
+        :rtype: :class:`~PySide.QtGui.QStandardItem`
+        """
+
+        if uid not in self._tree_data:
+            return None
+
+        return self._tree_data[uid]
+
+    def _load_cached_data(self):
+        """
+        Convenience wrapper from loading cached data from disk.
+
+        Handles logging cache load attempts/failures.
+
+        :returns: ``True`` if data loaded from disk, ``False`` otherwise.
+        """
+
+        # warn if the cache file does not exist
+        if not self._cache_path or not os.path.exists(self._cache_path):
+            logger.debug(
+                "Data cache file does not exist on disk.\n"
+                "Looking here: %s" % (self._cache_path,)
+            )
+            return False
+
+        logger.debug(
+            "Now attempting cached data load from: %s ..." %
+            (self._cache_path,)
+        )
+
+        try:
+            time_before = time.time()
+            num_items = self._load_from_disk()
```

Comment:
it seems `load_from_disk` isn't called from any subclasses so I think it should be made private, e.g. `__load_from_disk`. It's confusing, normally we handle "private" methods with underscores but in the case of class inheritance, we should try to make as much as possible `__private` and lave the `_single_underscore` only for methods intended to be subclassed (and official part of the supported interface).